Exiled.CustomItems/CustomItems.cs

```diff
@@ -36,24 +43,53 @@ private CustomItems()
         public override void OnEnabled()
         {
             roundHandler = new RoundHandler();
+            serverHandler = new ServerHandler();
 
             Events.Handlers.Server.RoundStarted += roundHandler.OnRoundStarted;
+            Events.Handlers.Server.SendingRemoteAdminCommand += serverHandler.OnRemoteAdminCommand;
             harmony = new Harmony($"com.{nameof(CustomItems)}.galaxy119-{DateTime.Now.Ticks}");
             harmony.PatchAll();
 
+            handle = Timing.RunCoroutine(ShowCustomItemInNickname());
             base.OnEnabled();
         }
 
         /// <inheritdoc />
         public override void OnDisabled()
         {
+            Timing.KillCoroutines(handle);
             Events.Handlers.Server.RoundStarted -= roundHandler.OnRoundStarted;
+            Events.Handlers.Server.SendingRemoteAdminCommand -= serverHandler.OnRemoteAdminCommand;
             harmony.UnpatchAll();
 
             harmony = null;
             roundHandler = null;
+            serverHandler = null;
 
             base.OnDisabled();
         }
+
+        private IEnumerator<float> ShowCustomItemInNickname()
```

Comment:
Is this necessary? Could the fake nickname be updated everytime someone changes its role and become a spectator? 
In servers with 50 players, it would iterate 2500 times every 0.5 seconds.

Reply:
So, shouldn't it update when you have a custom "hacky" item in your hand?
It could be done with the `ChangingItem` event.

Reply:
Since I don't want everyone on the server to see their new name, only spectators, we would still need to regularly check if someone is/isn't a spectator anymore, and change the user's name/reset it.

Reply:
Yeah, we could check it with `ChangingItem` and `ChangingRole`, to remove the custom nickname to non-spectators.

Reply:
But what happens if someone changes to the custom item, then someone new dies/joins the server. Their fake nickname would not be updated. 

This could be a problem if, for example, someone is using a gun that shoots through walls, and are reported by a player for 'hacking'. 
The joining gmod would not see the player's changed nickname.

Alternatively, if we make it so that anyone triggering ChangingRole with the new role being spectator, the above issue would be resolved, but we would have to foreach over every player when they change into something not spectator and re-sync each player's name.

`
switch (ev.NewRole == RoleType.Spectator)
{
    case true:
        foreach (Player player in Player.List)
            if (CustomItem.TryGet(player, out IEnumerable<Customitem> customItems))
                foreach (CustomItem item in customitems)
                    if (item.Check(player.CurrentItem))
                        ev.Player.SendFakeSyncVar(...)
            break;
    case false:
        foreach (Player player in Player.List)
            // reset player's nickname
        break;
}
`

I'm not sure that's all that much more efficient. It spreads out how often the foreach() is called, but will likely result in it being called *much* more frequently.

Reply:
> But what happens if someone changes to the custom item, then someone new dies/joins the server. Their fake nickname would not be updated.

That's why we should update the custom nickname when you change the custom item and when someone class gets changed to spectator (when you join, you get automatically changed to spectator).

The `ChangingRole` event will only update the custom nick for the joined player and not for all spectators (you don't need a foreach).
The `ChangingItem` event will change the custom nickname for all spectators (you only need a single foreach).

Reply:
In ChangingRole I need a foreach to go over a list of all the spawned players, check if they have custom items, then foreach over those custom items and check to see if any of them are their current weapon.

I also need to reset their nickname back if the player's NewRole isn't Spectator aswell, otherwise it defeats the purpose of only targeting spectators for the name change.

Reply:
> In ChangingRole I need a foreach to go over a list of all the spawned players, check if they have custom items, then foreach over those custom items and check to see if any of them are their current weapon.

Why? You can simply subscribe that event to `CustomItem` class and then change the nick of players who are using it, you may still need a foreach tho.
Same thing when that player spawns, but in that case you can reset every nicknames without the need to check for custom items.

Reply:
I'm in favor of @iopietro's event driven solution. Having a coroutine in the framework that iterates through all players every 0.5 seconds isn't a good idea.

Reply:
Very well, I will remove the coroutine and adapt an event driven solution and re-request your reviews shortly.